### ISSUE TITLE
perf(pcs): parallelize jagged partial_eval — searchsorted gather instead of serial per-column scatter

### ISSUE THREAD

**Issue by chokobole** [fractalyze/zorch#159]
### Pre-submission checklist

- [x] I have searched the existing issues and this is not a duplicate.

### Context / problem

`zorch/pcs/jagged/poly.py::partial_eval` materializes the indicator MLE `J̃(z_row, z_col, ·)` over the dense hypercube `(2^n_d,)` — the second factor of the outer Hadamard sumcheck — via

```python
def body(c, out):
    t_c = prefix_sums_int[c]
    contrib = col_eq[c] * jnp.where(row_indices < h, row_eq, 0)
    old = lax.dynamic_slice(out, (t_c,), (row_len,))
    return lax.dynamic_update_slice(out, old + contrib, (t_c,))
jax.lax.fori_loop(0, cfg.l_max, body, out)
```

The carried `out` plus overlapping write windows (`row_len` = padded max height, so consecutive columns' `[t_c, t_c + row_len)` regions overlap) make this a genuine RMW dependency chain: XLA lowers it as a sequential while thunk. On GPU that means **zero cross-column parallelism** — `l_max` serial iterations, each reading and writing `2^n_r` elements, plus per-iteration kernel-launch overhead:

```
bytes ≈ l_max × 2 × 2^n_r × elem_bytes   (fully serialized)
```

This runs once per `open()`. Absolute share at SP1-shaped layouts is unmeasured (GPU A/B is part of the plan below), but the serialization is structural — visible in the lowered HLO regardless of shape.

### Proposed solution

Invert scatter → gather: each output element finds its owning column instead of each column scattering its rows. All device-side, jittable, AOT-clean (no host numpy at trace time):

```python
i     = jnp.arange(1 << n_d, dtype=jnp.int32)
c     = jnp.searchsorted(prefix_sums_int, i, side="right") - 1   # owning column (prefix sums are monotone)
local = i - prefix_sums_int[c]
h     = prefix_sums_int[c + 1] - prefix_sums_int[c]
out   = jnp.where(local < h, col_eq[c] * row_eq[local], jnp.zeros([], dtype))
```

One elementwise kernel, fully parallel over `2^n_d` lanes. Zero-height padding columns are skipped naturally (they own no `i`); the tail region `i ≥ t_last` needs a clamp on `c` and masks to zero.

Following the repo convention for replacing a numerically-sensitive implementation, keep the fori_loop version as an oracle behind a kwarg (`use_gather=True` default flipped only after validation) with a bit-equal differential test.

Edge cases to pin in tests: zero-height columns, the `i ≥ t_last` tail, `local < row_len` (relies on the `h ≤ 2^n_r` layout invariant — assert or clamp), single-column and all-padding layouts.

### Alternatives considered

- **Keep the fori_loop** — rejected: the RMW carry serializes all `l_max` columns on GPU by construction; no tuning recovers parallelism.
- **Parallel scatter-add (`out.at[...].add(...)` over all columns at once)** — rejected: overlapping windows need atomic adds; XLA scatter-add on extension-field element types is unsupported/fragile, and it still materializes the per-column staging.
- **Host-precomputed segment-id constant array (heights are static per layout)** — rejected: bakes a `2^n_d`-sized constant into the graph/AOT artifact; `searchsorted` computes the same mapping on the fly for `log2(l_max)` compares per element.

### Why this approach

The gather form is the standard GPU answer to value-dependent scatter: every lane derives its own (column, local-row) coordinate from the monotone prefix sums, so coalescing is preserved on the output and the only irregular access is the small `prefix_sums_int`/`col_eq` tables (L2-resident, `l_max+1` entries). Auxiliary `int32` arrays (`i`, `c`, `local`) add `O(2^n_d)` bytes, but the fori version already moves `l_max × 2 × 2^n_r` — for realistic layouts the gather form wins on bytes *and* removes the serialization. It is also the same shape the eventual zkx custom emitter would want, so nothing is thrown away if codegen later absorbs it.

### Plan

- [x] `searchsorted`/gather implementation; fori_loop kept as oracle (in `testing/` per the reference-oracle convention, not a kwarg)
- [x] CPU differential test: bit-equal vs oracle across layouts (zero-height columns, tail region, single-column, all-padding)
- [x] GPU A/B at an SP1-shaped layout (waiting on free card); record numbers here
- [x] Flip the default to the gather path once A/B confirms (gather is the sole production path; the scatter form lives only in `testing/`)

### Acceptance criteria

- [x] Gather path is bit-identical to the fori oracle on all differential-test layouts (CPU)
- [x] GPU A/B numbers recorded on this issue (serial vs gather, same layout)
- [x] Path remains jittable/AOT-clean (no host numpy at trace time)

### References

- #42 (jagged PCS epic — parent)
- #121 (P3b: jagged opening — sumcheck reduction + stacked BaseFold open)
- #128 (stackable dense MLE, last touch of `poly.py`)
- Jagged Little Polynomial: ePrint 2025/917

### Area

none / unsure

### Parent issue

#42

### Milestone

jagged: Jagged PCS (ePrint 2025/917)

### Relevant knowledge

- Personal memo `zorch-proof-pipeline-tail-stages-bottleneck-classification-2026-06-05`: "jagged `partial_eval` = lax.fori_loop over ~hundreds of columns doing dynamic_slice/update_slice RMW — sequential per-column scatter on GPU; runs once per open; bytes ≈ l_max × 2 × 2^n_r × elem_bytes serialized. Inner assist rounds are clean (jit-once + reuse)." **(personal memo, not in team wiki)**
- Class prior from the zkx#521 encode work: round-loop/dispatch walls dominate proof-pipeline tail stages; this is the one site in the jagged prover where the lowered HLO is serial by construction.

**Comment by chokobole** [comment]
### A/B measurements (RTX 5090, idle GPU verified in-band pre/post 15 MiB; zkx feat/try-composite plugin)

| backend | shape | gather (new) | scatter (old) | ratio |
|---|---|---|---|---|
| GPU | l_max=64, n_r=10, n_d=17 | **0.098 ms** | 1.166 ms | **11.9× faster** |
| GPU | l_max=700, n_r=16, n_d=27 (SP1-ish) | **11.06 ms** | 15.61 ms | **1.41× faster** |
| CPU | l_max=700, n_r=16, n_d=27 | 4101 ms | 348 ms | 11.8× slower |

- The CPU regression is the accepted total-work trade (gather does O(2^n_d · log l_max) work vs scatter's O(l_max · 2^n_r); CPU executes serially so total work = time). Production proving is GPU; CPU tests use tiny layouts where the difference is sub-ms.
- The large-shape GPU win is modest because the gather kernel runs at ~195 GB/s effective vs the 2.1 GB output — likely multi-kernel materialization of the int intermediates; worth an HLO/fusion look only if this site shows up in a full-open profile. The scatter baseline scales linearly in l_max (serial while thunk), the gather logarithmically — at l_max≈1400 (multi-chip) the projected gap is ~2.8×.
- `method="compare_all"` searchsorted variant left unmeasured: at l_max=700 it is O(l_max) per lane (strictly worse); only relevant for tiny-l_max instances if those ever profile hot.

Bit-exactness: differential test (4 corner layouts byte-equal vs the scatter oracle) + the existing BP-oracle identity, both green on CPU.
